### PR TITLE
rubocops/cask: Check for correct stanza order within `on_*` blocks

### PR DESCRIPTION
**Pull request by issyl0**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

- Follow-up to #15211 to fix more TODOs but in a different cop.

-----

- Now that we detect correct stanza _grouping_ within `on_*` blocks in Casks, correct stanza _ordering_ in `on_*` blocks was the next logical step. For example, `url` has to come after `version` and `sha256` in an `on_macos` or `on_intel` block for consistency with the top-level stanza order we enforce elsewhere.
- Still not doing the nested `on_os` inside `on_arch`, that felt excessive for an edge case that isn't present in more than a few real Casks we have. I removed the test with that specific TODO. If anyone feels strongly I can look at it again, though.

### CODE REVIEW

File: Library/Homebrew/rubocops/cask/stanza_order.rb [hunk 1]
```diff
@@ -13,34 +13,48 @@ class StanzaOrder < Base
...
         MESSAGE = "`%<stanza>s` stanza out of order"
 
         def on_cask(cask_block)
           @cask_block = cask_block
-          add_offenses
+          add_offenses(toplevel_stanzas)
+
+          return unless (on_blocks = toplevel_stanzas.select { |s| ON_SYSTEM_METHODS.include?(s.stanza_name) }).any?
+
+          on_blocks.map(&:method_node).each do |on_block|
```
Comment (issyl0) [inline Library/Homebrew/rubocops/cask/stanza_order.rb:25]: We're doing this sort of thing in a bunch of cops, almost identically, so it should probably be moved somewhere central.

File: Library/Homebrew/test/rubocops/cask/stanza_order_spec.rb [hunk 2]
```diff
@@ -459,39 +459,22 @@
...
           on_arm do
-            url "https://foo.brew.sh/foo-arm.zip"
-            sha256 :no_check
             version :latest
-          end
+            sha256 :no_check
 
-          name "Foo"
+            url "https://foo.brew.sh/foo-arm.zip"
+          end
```
Comment (reitermarkus) [inline Library/Homebrew/test/rubocops/cask/stanza_order_spec.rb:477]: So this is only testing the order of `on_arm` and `on_intel`, and the one below tests only the order of inner stanzas. Should we have another test which tests both, as this did before, or merge these into one that tests both?

Comment (issyl0) [reply Library/Homebrew/test/rubocops/cask/stanza_order_spec.rb:477]: Ahhh, yes, this new test is important!

```
  1) RuboCop::Cop::Cask::StanzaOrder when the `on_os` stanzas and their contents are out of order autocorrects source
     Failure/Error: raise Parser::ClobberingError, "Parser::Source::TreeRewriter detected clobbering" if action == :raise
     
     Parser::ClobberingError:
       Parser::Source::TreeRewriter detected clobbering
     Shared Example Group: "autocorrects source" called from ./test/rubocops/cask/stanza_order_spec.rb:561
```

Comment (issyl0) [reply Library/Homebrew/test/rubocops/cask/stanza_order_spec.rb:477]: Hmmm, I see, it detects the toplevel stanzas (`on_mojave`) first, adds those offenses, then descends into the `on_mojave` block and detects those, then tries to correct them all but the ordering means that they all move about hence the "clobbering" from what it originally detected? Those are probably some bad words for explaining what's happening, but that's all I can come up with right now. No idea how to fix it, either.

I tried doing the _inner_ stanzas first but that doesn't help, but maybe I'm doing it wrong. 🤔

**Comment by issyl0** [comment]
The best I can come up with so far for the nested clobbering problem is fixing the inner stanzas first, then returning a log message if there are top level errors too asking the user to rerun the cop. That feels like awful UX.

**Comment by issyl0** [comment]
Putting this back in draft because I need to do more thinking about what's actually the problem here and the fixes for nested stanzas.

**Comment by issyl0** [comment]
OK, so the overwriting and infinite looping happens when determining the "correct stanza" for block inner stanzas:

```ruby
                correct_stanza_index = sorted_stanzas.index(stanza)
                correct_stanza = sorted_stanzas[correct_stanza_index]
```

Because a stanza named `:url` comes 15th in the list of sorted stanzas, whether or not it's a top-level URL or an inner URL (that can be in a different place since it's a different "kind", for want of a better term), so it tries to replace the 15th item, again, again, again...

I don't know how to not do that for inner stanzas. And I am quite confused and I feel like I'm going in circles. 😓

**Comment by issyl0** [comment]
OK, I thought I was getting somewhere last night with the reimagined approach after Markus' reply on Slack, but now I'm as confused as ever. Any ideas, folks? Is this worth it at this point?

File: Library/Homebrew/rubocops/cask/stanza_order.rb [hunk 3]
```diff
@@ -17,34 +17,38 @@ class StanzaOrder < Base
 
         def on_cask(cask_block)
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/cask/stanza_order.rb:18]: @issyl0, what I meant was that this cop shouldn't be run `on_cask`, but rather there should be something like `on_cask_stanza_block` instead, which matches `cask` and `on_*` blocks equally and this cop is then implemented in terms of `on_cask_stanza_block` rather than `on_cask`. So the cop itself doesn't have to loop through nested blocks.

No idea though if that makes a difference in how autocorrect behaves though.

Comment (issyl0) [reply Library/Homebrew/rubocops/cask/stanza_order.rb:18]: Unfortunately I still don't understand, sorry. How would it "treat cask and on blocks equally" when one is within the other? How to get to that state?

Comment (reitermarkus) [reply Library/Homebrew/rubocops/cask/stanza_order.rb:18]: I added a commit that adds `on_cask_stanza_block`, maybe that makes it clearer.

File: Library/Homebrew/rubocops/cask/mixin/cask_help.rb [hunk 4]
```diff
@@ -13,14 +13,20 @@ module CaskHelp
...
         def on_block(block_node)
           super if defined? super
-          return unless respond_to?(:on_cask)
-          return unless block_node.cask_block?
 
-          comments = processed_source.comments
-          cask_block = RuboCop::Cask::AST::CaskBlock.new(block_node, comments)
-          on_cask(cask_block)
+          if respond_to?(:on_cask_stanza_block) && (block_node.cask_block? || block_node.on_system_block?)
+            on_cask_stanza_block(block_node)
```
Comment (reitermarkus) [inline Library/Homebrew/rubocops/cask/mixin/cask_help.rb:22]: Here, `block_node` can now be a `cask` or a `on_*` block, i.e. they are treated equally without any sort of nested loop.

File: Library/Homebrew/rubocops/cask/stanza_order.rb (hunk 3 above)
Comment (issyl0) [reply Library/Homebrew/rubocops/cask/stanza_order.rb:18]: I thought it did help, but then I tried to use it and was confused again. I guess I don't understand how to make use of it and which bits it skips out having to do in the cop itself?

Comment (reitermarkus) [reply Library/Homebrew/rubocops/cask/stanza_order.rb:18]: > which bits it skips out having to do in the cop itself

Looping through nested blocks.

Comment (reitermarkus) [reply Library/Homebrew/rubocops/cask/stanza_order.rb:18]: > how to make use of it

I assume most of the needed methods are defined in `CaskBlock` which isn't used in this case.

**Comment by issyl0** [comment]
Right, even with a few days off I officially can't figure it out and it's reached the point where I get a sinking/anxious/useless feeling when I come back to it. Thanks so much for your continued help though, Markus! ❤️ If you'd like to continue with this then please do - it seems like you have an idea of what and how things could be achieved here and I wouldn't want your effort to be wasted too!

**Comment by reitermarkus** [comment]
@issyl0, I did it!

Also opened https://github.com/rubocop/rubocop/pull/11858 to fix an issue I encountered here.

**Comment by reitermarkus** [comment]
Also, another issue is the function `autocorrect_source` we use in our tests, which does not loop (to handle nested corrections). `rubocop-rspec` provides a `expect_correction` function which does this by default, which we should switch to.

**Comment by issyl0** [comment]
@reitermarkus 🤩 🤩 🤩

**Review by reitermarkus — APPROVED** [review]
